### PR TITLE
[WIP] SparseTensor Format

### PR DESCRIPTION
**Pull request by rusty1s**
This PR adds a universal `SparseTensor` class to `torch-sparse`. Due to lacking sparse matrix support in PyTorch, I decided to build my own class which allows to be easily extendable for all kind of use-cases. It builds upon the idea of providing fast conversions between COO, CSR and CSC formats by making use of internal caches. The following features are/should be supported:

- [x] Caching behaviour via `SparseStorage`
- [x] Optional values
- [x] Conversion utilities (dense, `torch.sparse_coo_tensor`, `scipy.sparse`)
- [x] Basic utility functions, *e.g.*, `to`, `detach`, `size`, `nnz`, ...
- [x] Advanced indexing via ints, slices and tensors following the Numpy/PyTorch notation, *e.g.*, `sparse_mat[10:20, index, ..., -1]`
- [x] Sparse-Dense Matrix Multiplication with custom reductions and broadcasting support
- [x] Sparse-Sparse Matrix Multiplication
- [x] Sparse Matrix reductions `sum`, `mean`, `prod`, `min` and `max` based on `scatter_*` and `segment_*`, *e.g.*, `sparse_mat.max(dim=0)` (`segment_max`) and `sparse_mat.max(dim=1)` (`scatter_max`)
- [x] Standard operators `add`, `sub`, `mul` and `div`, *e.g.*, `2 * sparse_mat`
- [x] `remove_diag` and `set_diag`
- [x] `coalesce`
- [x] Disabling caching behaviour (memory/computation trade-off) via `torch_sparse.set_cache_enabled(mode)`, `with torch_sparse.no_cache()` and `@torch_sparse.no_cache()`
- [ ] Docstrings
- [ ] Test Suite
- [x] Benchmark Suite

### CODE REVIEW

File: torch_sparse/sparse.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        assert len(sizes) == 2
+        self.__sparse_size__ = torch.Size(sizes)
+
+    def __is_sorted__(self):
+        idx1 = self.size(1) * index[0] + index[1]
+        idx2 = torch.cat([idx1.new_zeros(1), idx1[:-1]], dim=0)
+        return (idx1 >= idx2).all().item()
+
+    def __sort__(self):
+        idx = self.__sparse_size__(1) * self.__index__[0] + self.__index__[1]
```
Comment (tchaton) [inline torch_sparse/sparse.py]: Should be self.__sparse_size__[1]

Comment (rusty1s) [reply torch_sparse/sparse.py]: This is correct IMO, it corresponds to `num_cols * row + col` to achieve row-major ordering.

Comment (rusty1s) [reply torch_sparse/sparse.py]: I suggest you should be only looking at the latest commit. Should be fixed in somewhere in between.

File: torch_sparse/storage.py [hunk 2]
```diff
@@ -37,18 +39,34 @@ def __init__(self, row, col, value=None, sparse_size=None, rowptr=None,
...
         if colptr is None:
             ones = torch.ones_like(col) if ones is None else ones
             in_deg = scatter_add(ones, col, dim=0, dim_size=sparse_size[1])
             colptr = torch.cat([col.new_zeros(1), in_deg.cumsum(0)], dim=0)
+        else:
+            assert colptr.dtype == torch.long and colptr.device == col.device
+            assert colptr.dim() == 1 and colptr.size(0) == sparse_size[1] - 1
 
         if arg_csr_to_csc is None:
             idx = sparse_size[0] * col + row
```
Comment (tchaton) [inline torch_sparse/storage.py]: This array is already created beforehand, and argsorted if necessary. Could be interesting to reuse it if possible

File: cpu/arange_interleave.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#include <torch/extension.h>
+
+#include "compat.h"
+
+at::Tensor arange_interleave(at::Tensor start, at::Tensor repeat) {
```
Comment (tchaton) [inline cpu/arange_interleave.cpp]: Some comments could be nice there.

**Comment by klicperajo** [comment]
Broadcasting support would also be amazing! :)

**Comment by rusty1s** [comment]
Hi Johannes,

What ops should support broadcasting other than `matmul`?

**Comment by tchaton** [comment]
Hey @rusty1s, @klicperajo,

Maybe element-wise mul could be interesting too.

Best,
T.C

**Comment by klicperajo** [comment]
Having broadcasting for all the standard ops (add, sub, mul, div, max, min, mean, logsumexp) would be great! I often need these in a sparse manner, i.e. where operations only apply to non-zero entries. For example, add x_i to every non-zero value of matrix W in row i. I'm not sure how to best implement this API-wise, since these are not "normal" adding operations. But it seems like there should be a better way of broadcasting than with a full `index_select`.

**Comment by mpariente** [comment]
This looks great ! 
On the readme this is not really advertised, should we use this `SparseTensor` directly?

**Comment by rusty1s** [comment]
Yes, it's true that it's not advertised yet, sorry! I haven't got the time to write the docs yet. However, it's fully usable and most functions should be clear by looking at the function headers.

**Comment by mpariente** [comment]
Thanks for the answer, I'll have a look today. 
Second question: now that you have your own sparse tensor format, do you plan to add support to batched operations? I'd be very interested in batch sparse matrix dense vector product. 
Or, do you have pointers for where the minibatching is done in `pytorch_geometric` please?

**Comment by rusty1s** [comment]
The current format allows to apply sparse dense matrix multiplication where the dense matrix can be batched, *e.g.* `[num_nodes, num_nodes] @ [batch_size, num_nodes, num_features]`. For batching sparse-matrix multiplications, you can convert your sparse matrix to a diagonal stacked sparse matrix via `torch_sparse.cat([mat1, mat2], dim=(0, 1))`.

**Comment by mpariente** [comment]
Thanks again, I will try !
